File: lib/agent/research.ts
// Deep-analysis pipeline: perception -> multi-TF indicators -> news -> AI verdict -> persist
// Importable from Node.js scripts (no next/* imports).

import type { Perception } from './perception'
import type { AgentAnalysis, AgentVerdict } from './memory'
import { memory } from './memory'
import { buildSystemPrompt } from './system-prompt'
import { ema, sma, atr as calcAtr, rsi, adx } from './triggers'
import { createOpenAIClient, OPENROUTER_MODEL } from '../openrouter-client'
import { hlCall, fetchHLCandles, fetchAccountState } from '../hl-client'
import type { Candle, HLClearinghouseState, HLSpotClearinghouseState } from '../types'
import { readAgentConfig as readConfig } from './config-store'
import * as crypto from 'crypto'

interface IndicatorSnapshot {
  ema8: number | null
  ema21: number | null
  slopeUp: boolean | null
  rsi14: number | null
  atr14: number | null
  adx14: number | null
  lastClose: number
  lastTime: number
}

interface NewsResult {
  title: string
  description: string
  url: string
  recency: number
}

async function fetchFundingRate(coin: string): Promise<string> {
  try {
    const raw = await hlCall<Array<{ fundingRate: string }>>({ type: 'fundingHistory', coin, startTime: Date.now() - 86_400_000 })
    if (Array.isArray(raw) && raw.length > 0) {
      const r = parseFloat(raw[raw.length - 1].fundingRate)
      if (isFinite(r)) return `${(r * 100).toFixed(4)}%/hr`
    }
  } catch { /* skip */ }
  return 'N/A'
}

function computeIndicators(candles: Candle[]): IndicatorSnapshot {
  const closes = candles.map(c => c.c)
  const ema8Arr = ema(closes, 8)
  const ema21Arr = ema(closes, 21)
  const lastEma8 = ema8Arr[ema8Arr.length - 1]
  const lastEma21 = ema21Arr[ema21Arr.length - 1]

  let slopeUp: boolean | null = null
  if (isFinite(lastEma8) && ema8Arr.length >= 3 && isFinite(ema8Arr[ema8Arr.length - 3])) {
    slopeUp = lastEma8 > ema8Arr[ema8Arr.length - 3]
  }

  const rsi14Arr = rsi(candles, 14)
  const atr14Arr = calcAtr(candles, 14)
  const adx14Arr = adx(candles, 14)

  return {
    ema8: isFinite(lastEma8) ? lastEma8 : null,
    ema21: isFinite(lastEma21) ? lastEma21 : null,
    slopeUp,
    rsi14: isFinite(rsi14Arr[rsi14Arr.length - 1]) ? rsi14Arr[rsi14Arr.length - 1] : null,
    atr14: isFinite(atr14Arr[atr14Arr.length - 1]) ? atr14Arr[atr14Arr.length - 1] : null,
    adx14: isFinite(adx14Arr[adx14Arr.length - 1]) ? adx14Arr[adx14Arr.length - 1] : null,
    lastClose: closes[closes.length - 1] ?? 0,
    lastTime: candles[candles.length - 1]?.t ?? 0,
  }
}

async function fetchNews(coin: string): Promise<NewsResult[]> {
  const apiKey = process.env.BRAVE_API_KEY
  if (!apiKey) return []

  try {
    const query = encodeURIComponent(`${coin} crypto news today`)
    const res = await fetch(`https://api.search.brave.com/res/v1/web/search?q=${query}&count=3`, {
      headers: { 'X-Subscription-Token': apiKey, Accept: 'application/json' },
    })
    if (!res.ok) return []
    const data = await res.json() as { web?: { results?: Array<{ title: string; description: string; url: string; page_age?: string }> } }
    const results = (data.web?.results ?? []).map(r => ({
      title: r.title,
      description: r.description,
      url: r.url,
      recency: r.page_age ? Math.round((Date.now() - new Date(r.page_age).getTime()) / 3_600_000) : 0,
    }))
    return results.slice(0, 3)
  } catch {
    return []
  }
}

function buildUserMessage(
  coin: string,
  perception: Perception,
  tf1h: IndicatorSnapshot,
  tf4h: IndicatorSnapshot,
  tf1d: IndicatorSnapshot,
  fundingRate: string,
  newsHeadlines: NewsResult[],
  equity: number,
  openPositions: Array<{ coin: string; side: string; sizeUSD: number }>,
  mode: string,
): string {
  const triggerSummary = perception.triggers
    .filter(t => t.fired)
    .map(t => `${t.name}: ${t.reason}`)
    .join(', ') || 'no triggers fired'

  const indicatorBlock = (label: string, snap: IndicatorSnapshot): string => {
    const parts: string[] = []
    if (snap.ema8 !== null && snap.ema21 !== null) {
      parts.push(`EMA8=${snap.ema8.toFixed(4)}, EMA21=${snap.ema21.toFixed(4)}, ${snap.ema8 > snap.ema21 ? 'bullish' : 'bearish'}`)
    }
    if (snap.slopeUp !== null) parts.push(`EMA8 slope: ${snap.slopeUp ? 'rising' : 'falling'}`)
    if (snap.rsi14 !== null) parts.push(`RSI(14)=${snap.rsi14.toFixed(1)}`)
    if (snap.atr14 !== null) parts.push(`ATR(14)=${snap.atr14.toFixed(4)}`)
    if (snap.adx14 !== null) parts.push(`ADX(14)=${snap.adx14.toFixed(1)}`)
    parts.push(`last close=${snap.lastClose.toFixed(4)}`)
    return `${label}: ${parts.join(' | ')}`
  }

  const newsBlock = newsHeadlines.length > 0
    ? `News (top 3):\n${newsHeadlines.map(n => `- [${n.recency}h ago] ${n.title} — ${n.description}`).join('\n')}`
    : 'News: no headlines (BRAVE_API_KEY not set or empty results)'

  const positionBlock = openPositions.length > 0
    ? `Open positions: ${openPositions.map(p => `${p.coin} ${p.side} $${p.sizeUSD.toFixed(0)}`).join(', ')}`
    : 'Open positions: none'

  return [
    `Candidate: ${coin} (HL ${perception.type}-PERP)`,
    `Current mid: $${perception.mid.toFixed(4)}`,
    `Perception score: ${perception.compositeScore}/100`,
    `Fired triggers: ${triggerSummary}`,
    '',
    'Market context (multi-timeframe):',
    indicatorBlock('1h', tf1h),
    indicatorBlock('4h', tf4h),
    indicatorBlock('1d', tf1d),
    '',
    `Funding rate (latest): ${fundingRate}`,
    `Equity: $${equity.toFixed(2)}`,
    positionBlock,
    '',
    newsBlock,
    '',
    `Mode: ${mode} — ${mode === 'LIVE' ? 'your verdict will execute against real funds' : 'analysis only, no execution'}`,
    '',
    'Respond with 3-5 bullet points of reasoning, then output your decision as VALID JSON on the very last line:',
    '{"verdict":"PASS"|"LONG"|"SHORT"|"CLOSE","confidence":0.0-1.0,"side":"long"|"short"|null,"entryPx":number,"stopPx":number,"tpPx":number,"reasoning":"brief"}',
    'Nothing after the JSON.',
  ].join('\n')
}

async function callAI(systemPrompt: string, userMessage: string): Promise<string> {
  const client = createOpenAIClient()
  const response = await client.chat.completions.create({
    model: OPENROUTER_MODEL,
    messages: [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userMessage },
    ],
    stream: false,
    max_tokens: 1024,  // trimmed: 2-3 bullets + JSON only
    temperature: 0.1,
  })
  return response.choices[0]?.message?.content ?? ''
}

function parseVerdict(aiText: string, coin: string, perception: Perception): {
  verdict: AgentVerdict; confidence: number; side: 'long' | 'short' | null;
  entryPx: number; stopPx: number; tpPx: number; reasoning: string;
} {
  let verdict: AgentVerdict = 'PASS'
  let confidence = 0
  let side: 'long' | 'short' | null = null
  let entryPx = perception.mid
  let stopPx = 0
  let tpPx = 0
  let reasoning = aiText.trim()

  const lines = aiText.trim().split('\n')
  let jsonStr = ''
  for (let i = lines.length - 1; i >= 0; i--) {
    const line = lines[i].trim()
    if (line.startsWith('{') && line.includes('verdict') && line.endsWith('}')) {
      jsonStr = line
      break
    }
  }

  if (!jsonStr) {
    const match = aiText.match(/\{[^{}]*"verdict"[^{}]*\}/)
    if (match) jsonStr = match[0]
  }

  if (jsonStr) {
    try {
      const cleaned = jsonStr.replace(/```json?\s*/g, '').replace(/```\s*/g, '').trim()
      const parsed = JSON.parse(cleaned) as Record<string, unknown>

      const raw = String(parsed.verdict ?? '').toUpperCase()
      if (raw === 'LONG') verdict = 'LONG'
      else if (raw === 'SHORT') verdict = 'SHORT'
      else if (raw === 'CLOSE') verdict = 'CLOSE'

      confidence = typeof parsed.confidence === 'number' ? parsed.confidence : 0
      side = parsed.side === 'long' ? 'long' : parsed.side === 'short' ? 'short' : null
      entryPx = typeof parsed.entryPx === 'number' ? parsed.entryPx : perception.mid
      stopPx = typeof parsed.stopPx === 'number' ? parsed.stopPx : 0
      tpPx = typeof parsed.tpPx === 'number' ? parsed.tpPx : 0
      reasoning = typeof parsed.reasoning === 'string' ? parsed.reasoning : aiText.slice(0, 500)
    } catch {
      const firstLine = aiText.trim().split('\n')[0] ?? ''
      if (/LONG/i.test(firstLine)) verdict = 'LONG'
      else if (/SHORT/i.test(firstLine)) verdict = 'SHORT'
      else if (/CLOSE/i.test(firstLine)) verdict = 'CLOSE'
    }
  }

  return { verdict, confidence, side, entryPx, stopPx, tpPx, reasoning }
}

export async function research(coin: string, perception: Perception): Promise<AgentAnalysis> {
  try {
    const [c1h, c4h, c1d, fundingRaw] = await Promise.all([
      fetchHLCandles(coin, '1h', 100).catch(() => []),
      fetchHLCandles(coin, '4h', 100).catch(() => []),
      fetchHLCandles(coin, '1d', 60).catch(() => []),
      fetchFundingRate(coin),
    ])

    const tf1h = computeIndicators(c1h)
    const tf4h = computeIndicators(c4h)
    const tf1d = computeIndicators(c1d)

    // Skip news fetch — saves Brave API calls + ~400 tokens/call. Technical signals are the edge.
    const news: NewsResult[] = []

    const config = await readConfig()
    const mode = (config.mode as string) || 'OFF'

    let equity = 0
    let openPositions: Array<{ coin: string; side: string; sizeUSD: number }> = []
    const user = process.env.HYPERLIQUID_MASTER_ADDRESS || process.env.HYPERLIQUID_WALLET_ADDRESS || ''

    // Fetch perp equity directly (not via hl-client wrapper)
    const perpRes = await fetch(`https://api.hyperliquid.xyz/info`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ type: 'clearinghouseState', user }),
    })
    if (perpRes.ok) {
      const perp = await perpRes.json() as {
        marginSummary?: { accountValue: string; totalNtlPos: string }
        assetPositions?: Array<{ position: { coin: string; szi: string; entryPx: string; positionValue?: string } }>
      }
      equity = parseFloat(perp.marginSummary?.accountValue ?? '0')

      // On unified accounts, perp shows $0 — use spot balance as equity
      if (equity === 0) {
        const spotRes = await fetch(`https://api.hyperliquid.xyz/info`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ type: 'spotClearinghouseState', user }),
        })
        if (spotRes.ok) {
          const spot = await spotRes.json() as {
            balances?: Array<{ coin: string; total: string }>
          }
          equity = (spot.balances ?? [])
            .filter(b => ['USDC', 'USDT', 'USD'].includes(b.coin))
            .reduce((sum, b) => sum + parseFloat(b.total), 0)
        }
      }
      // Sync to memory so other routes see it
      memory.updateEquity(equity)

      openPositions = (perp.assetPositions ?? [])
        .filter(p => parseFloat(p.position.szi) !== 0)
        .map(p => {
          const szi = parseFloat(p.position.szi)
          const posVal = parseFloat(p.position.positionValue ?? '0')
          const sizeUSD = isFinite(posVal) && posVal > 0
            ? posVal
            : Math.abs(szi) * parseFloat(p.position.entryPx)
          return {
            coin: p.position.coin,
            side: szi > 0 ? 'long' : 'short',
            sizeUSD,
          }
        })
    }

    const wr = memory.getWinRate()

    const systemPrompt = buildSystemPrompt({ mode: mode as 'OFF' | 'LIVE', winRate: wr.rate, recentTrades: wr.total })
    const userMessage = buildUserMessage(coin, perception, tf1h, tf4h, tf1d, fundingRaw, news, equity, openPositions, mode)

    const aiText = await callAI(systemPrompt, userMessage)
    const parsed = parseVerdict(aiText, coin, perception)

    const analysis: AgentAnalysis = {
      id: crypto.randomUUID(),
      perceptionId: memory.getRecentPerceptions(1)[0]?.id ?? 'unknown',
      coin,
      verdict: parsed.verdict,
      confidence: parsed.confidence,
      side: parsed.side,
      entryPx: parsed.entryPx,
      stopPx: parsed.stopPx,
      tpPx: parsed.tpPx,
      reasoning: parsed.reasoning,
      newsContext: news.length > 0 ? news.map(n => n.title).join('; ') : 'no news',
      createdAt: Date.now(),
    }

    memory.recordAnalysis(analysis)
    return analysis
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    process.stderr.write(`[research] FAILED for ${coin}: ${msg}\n`)
    const fallback: AgentAnalysis = {
      id: crypto.randomUUID(),
      perceptionId: 'unknown',
      coin,
      verdict: 'PASS',
      confidence: 0,
      side: null,
      entryPx: perception.mid,
      stopPx: 0,
      tpPx: 0,
      reasoning: `Research failed: ${msg}`,
      createdAt: Date.now(),
    }
    memory.recordAnalysis(fallback)
    return fallback
  }
}

File: hermes_agent/agents/research.py
"""Deep-analysis pipeline: perception -> multi-timeframe indicators -> AI verdict -> persist."""

from __future__ import annotations

import asyncio
import json
import logging
import math
import os
import re
import time
import uuid
from typing import Any, Dict, List

import httpx

from hermes_agent.agents.config_store import read_agent_config
from hermes_agent.agents.memory import memory
from hermes_agent.agents.system_prompt import build_system_prompt
from hermes_agent.client.hl_client import (
    fetch_account_state,
    fetch_funding_history,
    fetch_hl_candles,
    resolve_user_address,
)
from hermes_agent.indicators.math import adx, atr, candle_val, ema, rsi
from hermes_agent.models.types import Candle

logger = logging.getLogger(__name__)


def _compute_indicators(candles: List[Candle]) -> Dict[str, Any]:
    """Compute EMA8/21, RSI, ATR, ADX for a set of candles."""
    if not candles:
        return {
            "ema8": None, "ema21": None, "slope_up": None,
            "rsi14": None, "atr14": None, "adx14": None,
            "last_close": 0, "last_time": 0,
        }

    closes = [candle_val(c, "c") for c in candles]

    if len(closes) < 21:
        return {
            "ema8": None, "ema21": None, "slope_up": None,
            "rsi14": None, "atr14": None, "adx14": None,
            "last_close": closes[-1],
            "last_time": candles[-1].t,
        }

    ema8_arr = ema(closes, 8)
    ema21_arr = ema(closes, 21)

    last_ema8 = ema8_arr[-1] if ema8_arr else None
    last_ema21 = ema21_arr[-1] if ema21_arr else None

    slope_up = None
    if last_ema8 is not None and len(ema8_arr) >= 3:
        slope_up = last_ema8 > ema8_arr[-3]

    rsi_arr = rsi(candles, 14)
    atr_arr = atr(candles, 14)
    adx_arr = adx(candles, 14)

    return {
        "ema8": last_ema8 if last_ema8 is not None and math.isfinite(last_ema8) else None,
        "ema21": last_ema21 if last_ema21 is not None and math.isfinite(last_ema21) else None,
        "slope_up": slope_up,
        "rsi14": rsi_arr[-1] if rsi_arr and math.isfinite(rsi_arr[-1]) else None,
        "atr14": atr_arr[-1] if atr_arr and math.isfinite(atr_arr[-1]) else None,
        "adx14": adx_arr[-1] if adx_arr and math.isfinite(adx_arr[-1]) else None,
        "last_close": closes[-1],
        "last_time": candles[-1].t,
    }


def _fetch_funding_rate(coin: str) -> str:
    """Latest hourly funding rate for a coin, or 'N/A' if unavailable."""
    start_time = int(time.time() * 1000) - 86_400_000
    history = fetch_funding_history(coin, start_time)
    if history:
        rate = float(history[-1].get("fundingRate", "0"))
        if math.isfinite(rate):
            return f"{rate * 100:.4f}%/hr"
    return "N/A"


def _build_user_message(
    coin: str,
    perception: Dict[str, Any],
    tf1h: Dict[str, Any],
    tf4h: Dict[str, Any],
    tf1d: Dict[str, Any],
    funding_rate: str,
    equity: float,
    open_positions: List[Dict[str, Any]],
    mode: str,
) -> str:
    """Build the user message passed to the LLM."""
    trigger_summary = (
        ", ".join(
            f"{t['name']}: {t['reason']}"
            for t in perception.get("triggers", [])
            if t.get("fired")
        )
        or "no triggers fired"
    )

    def _indicator_block(label: str, snap: Dict[str, Any]) -> str:
        parts = []
        if snap.get("ema8") is not None and snap.get("ema21") is not None:
            direction = "bullish" if snap["ema8"] > snap["ema21"] else "bearish"
            parts.append(
                f"EMA8={snap['ema8']:.4f}, EMA21={snap['ema21']:.4f}, {direction}"
            )
        if snap.get("slope_up") is not None:
            parts.append(f"EMA8 slope: {'rising' if snap['slope_up'] else 'falling'}")
        if snap.get("rsi14") is not None:
            parts.append(f"RSI(14)={snap['rsi14']:.1f}")
        if snap.get("atr14") is not None:
            parts.append(f"ATR(14)={snap['atr14']:.4f}")
        if snap.get("adx14") is not None:
            parts.append(f"ADX(14)={snap['adx14']:.1f}")
        parts.append(f"last close={snap.get('last_close', 0):.4f}")
        return f"{label}: {' | '.join(parts)}"

    position_block = (
        "Open positions: " + ", ".join(
            f"{p['coin']} {p['side']} ${p.get('size_usd', 0):.0f}"
            for p in open_positions
        )
        if open_positions
        else "Open positions: none"
    )

    return "\n".join([
        f"Candidate: {coin} (HL {perception.get('type', 'perp')}-PERP)",
        f"Current mid: ${perception.get('mid', 0):.4f}",
        f"Perception score: {perception.get('composite_score', 0)}/100",
        f"Fired triggers: {trigger_summary}",
        "",
        "Market context (multi-timeframe):",
        _indicator_block("1h", tf1h),
        _indicator_block("4h", tf4h),
        _indicator_block("1d", tf1d),
        "",
        f"Funding rate (latest): {funding_rate}",
        f"Equity: ${equity:.2f}",
        position_block,
        "",
        f"Mode: {mode} — {'your verdict will execute against real funds' if mode == 'LIVE' else 'analysis only, no execution'}",
        "",
        'Respond with 3-5 bullet points of reasoning, then output your decision as VALID JSON on the very last line:',
        '{"verdict":"PASS"|"LONG"|"SHORT"|"CLOSE","confidence":0.0-1.0,"side":"long"|"short"|"null","entryPx":number,"stopPx":number,"tpPx":number,"reasoning":"brief"}',
        "Nothing after the JSON.",
    ])


def _call_ai(system_prompt: str, user_message: str) -> str:
    """Call the OpenRouter LLM API (runs the async client in a fresh event loop)."""
    openrouter_key = os.environ.get("OPENROUTER_API_KEY", "")
    model = os.environ.get("OPENROUTER_MODEL", "x-ai/grok-4.3")

    if not openrouter_key:
        logger.warning("[research] OPENROUTER_API_KEY not set — returning empty response")
        return ""

    loop = asyncio.new_event_loop()
    try:
        return loop.run_until_complete(_async_do_call(openrouter_key, model, system_prompt, user_message))
    finally:
        loop.close()


async def _async_do_call(
    openrouter_key: str,
    model: str,
    system_prompt: str,
    user_message: str,
) -> str:
    """Async POST to the OpenRouter chat-completions endpoint."""
    async with httpx.AsyncClient(timeout=httpx.Timeout(120.0)) as client:
        resp = await client.post(
            "https://openrouter.ai/api/v1/chat/completions",
            json={
                "model": model,
                "messages": [
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_message},
                ],
                "stream": False,
                "max_tokens": 1024,
                "temperature": 0.1,
            },
            headers={"Authorization": f"Bearer {openrouter_key}"},
        )
        if resp.is_success:
            data = resp.json()
            choices = data.get("choices", [])
            if choices:
                return choices[0].get("message", {}).get("content", "")
    return ""


def parse_verdict(
    ai_text: str,
    coin: str,
    perception: Dict[str, Any],
) -> Dict[str, Any]:
    """Parse the AI response: JSON on the last line, with a regex fallback."""
    if not ai_text:
        ai_text = ""

    verdict = "PASS"
    confidence = 0.0
    side = None
    entry_px = perception.get("mid", 0)
    stop_px = 0.0
    tp_px = 0.0
    reasoning = ai_text.strip()

    lines = ai_text.strip().split("\n")

    # Find JSON on the last line
    json_str = ""
    for line in reversed(lines):
        line = line.strip()
        if line.startswith("{") and "verdict" in line and line.endswith("}"):
            json_str = line
            break

    # Fallback: regex match
    if not json_str:
        match = re.search(r'\{[^{}]*"verdict"[^{}]*\}', ai_text)
        if match:
            json_str = match.group(0)

    if json_str:
        try:
            cleaned = re.sub(r'```json\s*|```', '', json_str).strip()
            parsed = json.loads(cleaned)

            raw = str(parsed.get("verdict", "")).upper()
            if raw == "LONG":
                verdict = "LONG"
            elif raw == "SHORT":
                verdict = "SHORT"
            elif raw == "CLOSE":
                verdict = "CLOSE"

            confidence = parsed.get("confidence", 0)
            side = parsed.get("side") if parsed.get("side") in ("long", "short") else None
            entry_px = parsed.get("entry_px") or parsed.get("entryPx", perception.get("mid", 0))
            stop_px = parsed.get("stop_px") or parsed.get("stopPx", 0)
            tp_px = parsed.get("tp_px") or parsed.get("tpPx", 0)
            reasoning = parsed.get("reasoning", ai_text[:500])
        except json.JSONDecodeError:
            first_line = lines[0] if lines else ""
            if re.search(r"LONG", first_line, re.IGNORECASE):
                verdict = "LONG"
            elif re.search(r"SHORT", first_line, re.IGNORECASE):
                verdict = "SHORT"
            elif re.search(r"CLOSE", first_line, re.IGNORECASE):
                verdict = "CLOSE"

    return {
        "verdict": verdict,
        "confidence": confidence,
        "side": side,
        "entry_px": entry_px,
        "stop_px": stop_px,
        "tp_px": tp_px,
        "reasoning": reasoning,
    }


def research(coin: str, perception: Dict[str, Any]) -> Dict[str, Any]:
    """Full AI research pipeline for a perception — returns an analysis dict."""
    c1h = fetch_hl_candles(coin, "1h", 100)
    c4h = fetch_hl_candles(coin, "4h", 100)
    c1d = fetch_hl_candles(coin, "1d", 60)

    funding_raw = _fetch_funding_rate(coin)

    if len(c4h) < 30:
        logger.warning(f"[research] thin 4h history for {coin}: only {len(c4h)} candles")

    tf1h = _compute_indicators(c1h)
    tf4h = _compute_indicators(c4h)
    tf1d = _compute_indicators(c1d)

    config = read_agent_config()
    mode = str(config.get("mode", "OFF"))

    equity = 0.0
    open_positions: List[Dict[str, Any]] = []
    user = resolve_user_address()

    if user:
        state = fetch_account_state(user)
        equity = float(state.get("equity", "0"))
        memory.update_equity(equity)

        open_positions = [
            {
                "coin": p.get("position", {}).get("coin", ""),
                "side": "long" if float(p.get("position", {}).get("szi", "0")) > 0 else "short",
                "size_usd": float(p.get("position", {}).get("positionValue", "0")) or (
                    abs(float(p.get("position", {}).get("szi", "0"))) *
                    float(p.get("position", {}).get("entryPx", "0"))
                ),
            }
            for p in (state.get("asset_positions") or [])
            if float(p.get("position", {}).get("szi", "0")) != 0
        ]

    wr = memory.get_win_rate()
    system_prompt = build_system_prompt(mode, wr.get("rate", 0), int(wr.get("total", 0)))
    user_message = _build_user_message(
        coin, perception, tf1h, tf4h, tf1d,
        funding_raw, equity, open_positions, mode,
    )

    ai_text = _call_ai(system_prompt, user_message)
    parsed = parse_verdict(ai_text, coin, perception)

    analysis = {
        "id": str(uuid.uuid4()),
        "perception_id": perception.get("id", "unknown"),
        "coin": coin,
        "verdict": parsed["verdict"],
        "confidence": parsed["confidence"],
        "side": parsed["side"],
        "entry_px": parsed["entry_px"],
        "stop_px": parsed["stop_px"],
        "tp_px": parsed["tp_px"],
        "reasoning": parsed["reasoning"],
        "news_context": "no news",
        "created_at": int(time.time() * 1000),
    }

    memory.record_analysis(analysis)
    return analysis
